### PR TITLE
kube-dns-anti-affinity: kube-dns never-co-located-in-the-same-node

### LINKED ISSUE #2693: Kube-dns pods shouldn't all be on one node
We are experiencing failures due to all kube-dns pods landing on a single node (2 of 2 pods in our case), and that node eventually suffers any issue (network outage, physical host issue) and then all DNS services stop which causes all of our applications to fail.

It would be a great improvement if kops supported some method to ensure that this doesn't happen, perhaps via daemonsets or affinities.

This is on kubernetes 1.5.3 and kops 1.5.3

### PR DESCRIPTION
**Pull request by StevenACoffman**
**What this PR does / why we need it**:

This is upstreaming the kubernetes/kops#2705 pull request by @jamesbucher that was originally against [kops](github.com/kubernetes/kops).
Please see kubernetes/kops#2705 for more details, including a lengthy discussion.

Briefly, given the constraints of how the system works today:

+ if you need multiple DNS pods primarily for availability, then requiredDuringSchedulingIgnoredDuringExecution makes sense because putting more than one DNS pod on the same node isn't useful
+ if you need multiple DNS pods primarily for performance, then
preferredDuringScheduling IgnoredDuringExecution makes sense because it will allow the DNS pods to schedule even if they can't be spread across nodes

**Which issue this PR fixes**

fixes kubernetes/kops#2693

**Release note**:

```release-note
Improve resilience by annotating kube-dns addon with podAntiAffinity to prefer scheduling on different nodes.
```

### CODE REVIEW

**Comment by StevenACoffman** [comment]
@justinsb @chrislovecnm I'm not familiar enough with the test suites to understand why this change would cause these errors:
```
W0909 20:42:43.501] error: context "e2e-f8n-agent-pr-93-0" does not exist
W0909 21:04:47.369] Error from server (InternalError): Internal error occurred: Authorization error (user=kube-apiserver, verb=get, resource=nodes, subresource=proxy)
W0909 21:04:47.635] error: context "e2e-f8n-agent-pr-93-0" does not exist
W0909 21:06:54.694] 2017/09/09 21:06:54 main.go:233: Something went wrong: error starting federation: error during ./federation/cluster/federation-up.sh: exit status 124
I0909 21:06:55.796] Starting pull-kubernetes-federation-e2e-gce-26305...
E0909 21:06:55.796] Command failed
I0909 21:06:55.797] process 20344 exited with code 1 after 39.8m
E0909 21:06:55.797] FAIL: pull-kubernetes-federation-e2e-gce
```

**Comment by bowei** [comment]
@csbell @kubernetes/sig-federation-misc for federation e2e failures

**Comment by MrHohn** [comment]
/retest

Ref https://github.com/kubernetes/kubernetes/issues/40063, https://github.com/kubernetes/kubernetes/issues/41125

cc @thockin

**Comment by chrislovecnm** [comment]
The main question that was raised on the PR into kops was if we use required or perferred. @thockin et al any thoughts?

**Comment by johanneswuerbach** [comment]
Any updates on this? We also just lost cluster dns service as all dns pods were located on the same node.

Anything I could help?

**Comment by bowei** [comment]
There is a discussion on the kops repo about using "preferred" vs "required". https://github.com/kubernetes/kops/pull/2705

It seems to me we should go with "preferred" for now as it is strictly better than what we have before and does not result in kube-dns instances not being scheduled due to constraints.

If the PR is updated to preferred, I can lgtm.

**Comment by StevenACoffman** [comment]
@bowei Done!

**Comment by bowei** [comment]
Can you also update kubeadm? Unfortunately, this needs to be maintained separately for now.

**Comment by StevenACoffman** [comment]
What the what now? The [kubeadm repo](https://github.com/kubernetes/kubeadm) has no equivalent files... so I'm not sure where that really lives.

**Comment by johanneswuerbach** [comment]
Assumably https://github.com/kubernetes/kubernetes/blob/49b4a514d802cc622f6b518c538366dd874d1150/cmd/kubeadm/app/phases/addons/dns/manifests.go

File: cluster/addons/dns/kube-dns.yaml.base [hunk 1]
```diff
@@ -83,7 +83,38 @@ spec:
         k8s-app: kube-dns
       annotations:
         scheduler.alpha.kubernetes.io/critical-pod: ''
+        # For 1.6, we keep the old tolerations in case of a downgrade to 1.5
```
Comment (MrHohn) [inline cluster/addons/dns/kube-dns.yaml.base]: Could we get rid of these annotations? https://github.com/kubernetes/kubernetes/pull/52193#discussion_r139030187

Comment (StevenACoffman) [reply cluster/addons/dns/kube-dns.yaml.base]: So delete lines 86 -107 here and the equivalents?

Comment (MrHohn) [reply cluster/addons/dns/kube-dns.yaml.base]: Yep, that sounds correct to me.

Comment (StevenACoffman) [reply cluster/addons/dns/kube-dns.yaml.base]: Are you asking me? I have no way of telling if it is at all common to have the newer kube-dns used by older kubernetes versions, so I have no opinion.

Comment (MrHohn) [reply cluster/addons/dns/kube-dns.yaml.base]: IMHO we shouldn't be carrying the deprecated alpha annotations on the upstream kube-dns manifest. Specific k8s deployment could have its own addon configuration to fit its use case, but that should be maintained separately. So I'd prefer we delete them here.

**Comment by MrHohn** [comment]
From https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/52193/pull-kubernetes-e2e-gce-etcd3/58726/artifacts/e2e-58726-master/kube-addon-manager.log:
```
error: error validating "/etc/kubernetes/addons/dns/kube-dns.yaml": error validating data: [field spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].weight for v1.WeightedPodAffinityTerm is required, field spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm for v1.WeightedPodAffinityTerm is required, found invalid field labelSelector for v1.WeightedPodAffinityTerm, found invalid field topologyKey for v1.WeightedPodAffinityTerm]; if you choose to ignore these errors, turn validation off with --validate=false
```

Seems the sub-fields of `preferredDuringSchedulingIgnoredDuringExecution` isn't constructed correctly.

**Comment by StevenACoffman** [comment]
I removed the old annotations, but if I somehow munged the sub-fields I'll need to look at it in the morning.

**Comment by StevenACoffman** [comment]
Other than finding and replacing "required" to "preferred" these look identical to the original downstream (kops) pull request so I'm not seeing why they are failing now.

**Comment by StevenACoffman** [comment]
Based on @MrHohn I assume that the format is incorrect, and I don't know why it passed before. However, looking at some other examples, I am guessing this is better?:
```
          preferredDuringSchedulingIgnoredDuringExecution:
          - weight: 1
            podAffinityTerm:
              labelSelector:
                matchExpressions:
                  - key: k8s-app
                    operator: In
                        values: ["kube-dns"]
              topologyKey: kubernetes.io/hostname
```

**Comment by MrHohn** [comment]
@StevenACoffman Yeah, seems `requiredDuringSchedulingIgnoredDuringExecution` and `preferredDuringSchedulingIgnoredDuringExecution` have slightly different schemes. What you posted looks correct to me, just that I think we should set `weight` to the highest value (100).

Ref: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/podaffinity.md

**Comment by StevenACoffman** [comment]
Thanks to @MrHohn I made a separate commit for the new scheme. Waiting for the tests now.

File: cluster/addons/dns/kube-dns.yaml.base [hunk 2]
```diff
@@ -84,6 +84,17 @@ spec:
       annotations:
         scheduler.alpha.kubernetes.io/critical-pod: ''
     spec:
+      affinity:
+        podAntiAffinity:
+          preferredDuringSchedulingIgnoredDuringExecution:
+          - weight: 1
```
Comment (MrHohn) [inline cluster/addons/dns/kube-dns.yaml.base]: We should use 100 here (the highest value).

Comment (StevenACoffman) [reply cluster/addons/dns/kube-dns.yaml.base]: Ah! Sorry, I forgot. I corrected it.

**Comment by StevenACoffman** [comment]
Looks like the tests passed, and the weight is corrected. How's this look to you now @MrHohn and @bowei ?

**Comment by MrHohn** [comment]
Thanks!
/lgtm

Could you update release note as well? Ref [PULL_REQUEST_TEMPLATE](https://github.com/kubernetes/kubernetes/blob/4a01f44b7378fb0c58450898e1bca30ea8a5158a/.github/PULL_REQUEST_TEMPLATE.md).

**Comment by evanj** [comment]
YAY! Thanks! I just noticed that one of our clusters was running both its instances of kube-dns on a single machine today, then found this issue while searching around for bugs or mentions. I was going to attempt this, but this is even better!

**Comment by luxas** [comment]
This got reverted now, but please do keep kubeadm in sync at all times. The code lives in `cmd/kubeadm` as @johanneswuerbach pointed out (thanks!)

**Comment by jonastl** [comment]
1. For what reason was this reverted?
2. Was #54164 referring to an actual production cluster, or a "toy test" setup?
3. What is the current plan for ensuring the intended behavior of this PR?

I just opened a [ticket](https://github.com/kubernetes/website/issues/6456) on this very issue (in the wrong project unfortunately), as result of the described behavior causing problems in one of our production clusters.

We have three small control nodes which service critical infrastructure pods, and a slew of pre-emptive nodes (on GKE). Without this patch we end up with poor scheduling decisions, where the scheduler stacks kube-dns pods on top of each other on the same nodes, causing our own workload pods to fail scheduling due to kube-dns having stolen the Allocatable resources.

We had hoped to be able to use the merged patch with node anti-affinity, to keep kube-dns away from the preemptive worker nodes, to avoid the problem faced and described in #41125 (spotty DNS service).

Without this patch we'll end up with dozens of kube-dns pods (created by kube-dns-autoscaler), all  stacked onto 3 x 2 core machines, stealing all allocatable CPU and preventing our own critical backplane pods from being scheduled.

Right now it seems our only option is to buy more machines for the sole purpose of hosting kube-dns. The pod(s) must run somewhere in the cluster, but we don't want them on pre-emptive nodes, and we don't want more than at most one on each control node. Since the reverted merge prevents us from telling the scheduler "not more than one per node!", it seems to me we'll need a new set of machines dedicated to kube-dns hosting. This will increase our cost and there will be more moving parts to manage (additional node group).

**Comment by bowei** [comment]
Unfortunately the scheduling feature (anti-affinity) does not currently scale very well, it cannot be used for a system service such as this. When the scheduling team improves the performance of the feature, we can re-examine the scheduler constraint...

**Comment by jonastl** [comment]
Bowei, do you have a proposal / suggestion for how to cope with kubedns until such a fix is out?
Is a dedicated kube-dns "server farm" the only option?

**Comment by bowei** [comment]
It possible to run a customized version of kube-dns on the cluster:  

NOTE: after following these steps, you will be responsible for maintaining the configuration of the DNS system. Changes can always be reverted to return to the original configuration.

* Download your current cluster kube-dns yaml: 
  * `kubectl get -n kube-system -o yaml deployment/kube-dns > my-kube-dns.yaml`
  * Modify the YAML with your custom settings.
  * Change the name of the deployment in the metadata section to my-kube-dns.
  * Remove the label `kubernetes.io/cluster-service: "true"` from my-kube-dns.yaml. (The label is used by the Kubernetes addon manager to distinguish user resources from those managed by the Kubernetes system itself)
  * Make sure the cluster domain is set properly in my-kube-dns.yaml. (Search for cluster.local)
* Scale the dns autoscaler replicas to 0. You can use kubectl scale --namespace kube-system deploy/kube-dns-autoscaler --replicas 0.
* The next steps will disrupt DNS temporarily in the cluster:
* Create the deployment my-kube-dns.yaml as constructed above. ClusterFirst requests will be sent to my-kube-dns as well as kube-dns because they have the same application label.
* Scale the existing `kube-system:kube-dns` replicas to 0 to remove traffic from kube-dns. You can do this by setting the DNS auto scaler to have 0 replicas. 
* Scale up my-kube-dns by setting replicas to the appropriate value.

**Comment by jonastl** [comment]
Excellent. Just one question.
What about the label `addonmanager.kubernetes.io/mode: Reconcile`? 
I take it that should be removed as well.

**Comment by bowei** [comment]
Yes, anything related to the addon manager will need to be removed.

**Comment by jonastl** [comment]
Perfect, thanks!
This information will prove useful to anyone else coming here facing the same problem (and using your company as hosting provider).

**Comment by StevenACoffman** [comment]
> Unfortunately the scheduling feature (anti-affinity) does not currently scale very well, it cannot be used for a system service such as this. When the scheduling team improves the performance of the feature, we can re-examine the scheduler constraint...
@bowei Is there an issue tracking the anti-affinity performance/scaling problem we can link to this PR to remind us to revisit it?

**Comment by MrHohn** [comment]
> bowei Is there an issue tracking the anti-affinity performance/scaling problem we can link to this PR to remind us to revisit it?

I believe that issue is https://github.com/kubernetes/kubernetes/issues/54189.

**Pull request merged**